surfactant/sbomtypes/_sbom.py

```diff
@@ -113,3 +114,272 @@ def create_software(
+                    yUUID=rel.yUUID,
+                    relationship=rel.relationship,
+                ):
+                    print(f"DUPLICATE STAR RELATIONSHIP: {existing_rel}")
+                else:
+                    merged_sbom.starRelationships.append(rel)
+        if sbom_m.starRelationships:
+            for rel in self.starRelationships:
+                # rewrite UUIDs before doing the search
+                if rel.xUUID in uuid_updates:
+                    rel.xUUID = uuid_updates[rel.xUUID]
+                if rel.yUUID in uuid_updates:
+                    rel.yUUID = uuid_updates[rel.yUUID]
+                if existing_rel := self._find_star_relationship_entry(
+                    merged_sbom,
+                    xUUID=rel.xUUID,
+                    yUUID=rel.yUUID,
+                    relationship=rel.relationship,
+                ):
+                    print(f"DUPLICATE STAR RELATIONSHIP: {existing_rel}")
+                else:
+                    merged_sbom.starRelationships.append(rel)
+        return merged_sbom
+
+    def _find_systems_entry(self, sbom, uuid=None, name=None):
+        """Merge helper function to find and return
+        the matching system entry in the provided sbom.
+
+        Positional arguments:
+        sbom -- the merged sbom
+        uuid -- the uuid of the desired system entry
+        name -- the name of the desired system entry
+        """
+        for system in sbom.systems:
+            all_match = True
+            if uuid:
+                if system.UUID != uuid:
+                    all_match = False
+            if name:
+                if system.name != name:
+                    all_match = False
+            if all_match:
+                return system
+        return None
+
+    def _find_software_entry(self, sbom, uuid, sha256=None, md5=None, sha1=None):
+        """Merge helper function to find and return
+        the matching software entry in the provided sbom.
+
+        Positional arguments:
+        sbom -- the merged sbom
+        uuid -- the uuid of the desired software entry
+        sha256 -- the sha256 of the desired software entry
+        md5 -- the md5 of the desired software entry
+        sha1 -- the sha1 of the desired software entry
+        """
+        for sw in sbom.software:
+            match = False
+            # If we have hashes to check
+            if sha256 or md5 or sha1:
+                # Check if we have both sides of the comparison, then compare. At least one hash must match
+                if sw.sha256 and sha256:
+                    if sw.sha256 == sha256:
+                        match = True
+                if sw.md5 and md5:
+                    if sw.md5 == md5:
+                        match = True
+                if sw.sha1 and sha1:
+                    if sw.sha1 == sha1:
+                        match = True
+            # If no hashes to check, match by UUID
+            else:
+                if sw.UUID == uuid:
```

Comment:
Ah okay, could you clarify what the suggestion is here? Instead of checking for hashes vs checking for uuid, we should just check all of them at once?

Reply:
I think I was mostly checking that this function still behaves as it did before, even after setting a default value of `None` for the `uuid` argument.

From the perspective of a user, checking all of them at once and requiring all of the arguments that are not `None` to match in order to consider an entry found feels like it would be more intuitive, instead of one argument being a fallback for the others. Or I'd expect the "found" criteria to be `<UUIDs match> and <any hashes that are present in both entries must match>`. For the purposes of the merging functionality working as expected maybe there's something that requires matching by UUID as a fallback that I'm not seeing.

Reply:
Ah I see, so I was running into an issue where I had two sboms with different UUIDs for the same entry for "helics.tar.gz". Surfactant was then failing on the match since a UUID was provided, but the UUIDs didnt match, even though the hashes all matched. So here the logical check from my perspective is, check that the file hashes match because those will be most accurate. If no file hashes exist, check the UUID because if that matches, it is likely the same software entry.